glib/src/subclass/type_plugin.rs

```diff
@@ -0,0 +1,323 @@
+        &self,
+        _instance_type: Type,
+        _interface_type: Type,
+        _interface_info: &InterfaceInfo,
+    );
+    fn register_dynamic_enum(&self, _name: &str, _const_static_values: &EnumValue) -> Type;
+    fn register_dynamic_flags(&self, _name: &str, _const_static_values: &FlagsValue) -> Type;
+    fn register_dynamic_type(
+        &self,
+        _parent_type: Type,
+        _type_name: &str,
+        _type_info: &TypeInfo,
+        _flags: TypeFlags,
+    ) -> Type;
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{self as glib, prelude::TypePluginExt};
+
+    use super::*;
+
+    mod imp {
+        use std::ffi::CString;
+
+        use super::*;
+        use crate::object::ObjectType;
+
+        #[derive(Default)]
+        pub struct SimplePlugin {
+            type_info: std::cell::Cell<Option<TypeInfo>>,
+        }
+
+        #[crate::object_subclass]
+        impl ObjectSubclass for SimplePlugin {
+            const NAME: &'static str = "SimplePlugin";
+            type Type = super::SimplePlugin;
+            type Interfaces = (TypePlugin,);
+        }
+
+        impl ObjectImpl for SimplePlugin {}
+
+        impl TypePluginImpl for SimplePlugin {
+            fn use_plugin(&self) {
+                // register types on implementation load
+                SimplePluginType::on_implementation_load(self.obj().as_ref());
+            }
+
+            fn unuse_plugin(&self) {
+                // unregister types on implementation unload
+                SimplePluginType::on_implementation_unload(self.obj().as_ref());
+            }
+
+            fn complete_type_info(&self, _type_: Type) -> (TypeInfo, TypeValueTable) {
+                assert!(self.type_info.get().is_some());
+                // return type info
+                (self.type_info.get().unwrap(), TypeValueTable::default())
+            }
+        }
+
+        impl TypePluginRegisterImpl for SimplePlugin {
+            fn add_dynamic_interface(
+                &self,
+                _instance_type: Type,
+                _interface_type: Type,
+                _interface_info: &InterfaceInfo,
+            ) {
+                unimplemented!()
+            }
+
+            fn register_dynamic_enum(&self, _name: &str, _const_static_values: &EnumValue) -> Type {
+                unimplemented!()
+            }
+
+            fn register_dynamic_flags(
+                &self,
+                _name: &str,
+                _const_static_values: &FlagsValue,
+            ) -> Type {
+                unimplemented!()
+            }
+
+            fn register_dynamic_type(
+                &self,
+                parent_type: Type,
+                type_name: &str,
+                type_info: &TypeInfo,
+                flags: TypeFlags,
+            ) -> Type {
+                let type_ = unsafe {
+                    let type_name = CString::new(type_name).unwrap();
+                    let mut g_type = gobject_ffi::g_type_from_name(type_name.as_ptr());
```

Comment:
```suggestion
                    let mut g_type = gobject_ffi::g_type_from_name(type_name.to_glib_none().0);
```

or

```suggestion
                    let mut g_type = type_name.run_with_gstr(|type_name| gobject_ffi::g_type_from_name(type_name.as_ptr()));
```

also elsewhere